// angular2-cool-http/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var core_1 = require("@angular/core");
var http_1 = require("@angular/http");
var cool_http_service_1 = require("./src/cool-http.service");
var cool_http_service_2 = require("./src/cool-http.service");
exports.CoolHttp = cool_http_service_2.CoolHttp;
var http_header_model_1 = require("./src/http-header.model");
exports.HttpHeader = http_header_model_1.HttpHeader;
var CoolHttpModule = (function () {
    function CoolHttpModule() {
    }
    return CoolHttpModule;
}());
CoolHttpModule = __decorate([
    core_1.NgModule({
        exports: [],
        imports: [http_1.HttpModule],
        providers: [cool_http_service_1.CoolHttp]
    })
], CoolHttpModule);
exports.CoolHttpModule = CoolHttpModule;
